nexus/tests/integration_tests/sp_updater.rs

```diff
@@ -0,0 +1,445 @@
+    ));
+
+    // Configure and instantiate an `SpUpdater`.
+    let sp_type = SpType::Sled;
+    let sp_slot = 0;
+    let update_id = Uuid::new_v4();
+    let hubris_archive = make_fake_sp_image(SIM_GIMLET_BOARD);
+
+    let sp_updater = SpUpdater::new(
+        sp_type,
+        sp_slot,
+        update_id,
+        hubris_archive.clone(),
+        &mgstestctx.logctx.log,
+    );
+
+    // Run the update.
+    sp_updater.update([mgs_client]).await.expect("update failed");
+
+    // Ensure the SP received the complete update.
+    let last_update_image = mgstestctx.simrack.gimlets[sp_slot as usize]
+        .last_update_data()
+        .await
+        .expect("simulated SP did not receive an update");
+
+    let hubris_archive = RawHubrisArchive::from_vec(hubris_archive).unwrap();
+
+    assert_eq!(
+        hubris_archive.image.data.as_slice(),
+        &*last_update_image,
+        "simulated SP update contents (len {}) \
+         do not match test generated fake image (len {})",
+        last_update_image.len(),
+        hubris_archive.image.data.len()
+    );
+}
+
+#[tokio::test]
+async fn test_sp_updater_updates_switch() {
+    // Start MGS + Sim SP.
+    let (mgs_config, sp_sim_config) = mgs_setup::load_test_config();
+    let mgs_addr = SocketAddrV6::new(Ipv6Addr::LOCALHOST, 0, 0, 0);
+    let mgstestctx = mgs_setup::test_setup_with_config(
+        "test_sp_updater_updates_switch",
+        SpPort::One,
+        mgs_config,
+        &sp_sim_config,
+        Some(mgs_addr),
+    )
+    .await;
+
+    let mgs_listen_addr = mgstestctx
+        .server
+        .dropshot_server_for_address(mgs_addr)
+        .expect("missing dropshot server for localhost address")
+        .local_addr();
+    let mgs_client = Arc::new(gateway_client::Client::new(
+        &format!("http://{mgs_listen_addr}"),
+        mgstestctx.logctx.log.new(slog::o!("component" => "MgsClient")),
+    ));
+
+    let sp_type = SpType::Switch;
+    let sp_slot = 0;
+    let update_id = Uuid::new_v4();
+    let hubris_archive = make_fake_sp_image(SIM_SIDECAR_BOARD);
+
+    let sp_updater = SpUpdater::new(
+        sp_type,
+        sp_slot,
+        update_id,
+        hubris_archive.clone(),
+        &mgstestctx.logctx.log,
+    );
+
+    sp_updater.update([mgs_client]).await.expect("update failed");
+
+    let last_update_image = mgstestctx.simrack.sidecars[sp_slot as usize]
+        .last_update_data()
+        .await
+        .expect("simulated SP did not receive an update");
+
+    let hubris_archive = RawHubrisArchive::from_vec(hubris_archive).unwrap();
+
+    assert_eq!(
+        hubris_archive.image.data.as_slice(),
+        &*last_update_image,
+        "simulated SP update contents (len {}) \
+         do not match test generated fake image (len {})",
+        last_update_image.len(),
+        hubris_archive.image.data.len()
+    );
+}
+
+#[tokio::test]
+async fn test_sp_updater_remembers_successful_mgs_instance() {
```

Comment:
Nice test.  Is it worth one that makes sure that if an MGS instance fails partway through, then we do go back to the other one?  I could go either way -- we're going to need broader integration-level tests for stuff like that and it might not make sense to test that level of detail at this layer.

Reply:
Seems reasonable! I'll take a stab at it; happy to add it if it's relatively easy (which I think it will be).

Reply:
Added in 53fc4d8dc